File: abc/abc192/e.py
import heapq
import math

n, m, x, y = map(int, input().split())
x, y = x-1, y-1

path = [[] for _ in range(n)]

for i in range(m):
    a, b, t, k = map(int, input().split())
    a, b = a-1, b-1
    path[a].append((b, t, k))
    path[b].append((a, t, k))

min_time = [float("inf") for _ in range(n)]

heap = [(0, x)]


while len(heap):
    # print(heap)
    time, town = heapq.heappop(heap)
    # print("town:", town, time)
    if min_time[y] < time:
        break
    for next, t, k in path[town]:
        next_time = ((time - 1) // k + 1) * k + t
        if min_time[next] > next_time:
            # print("add", next, next_time)
            min_time[next] = next_time
            heapq.heappush(heap, (next_time, next))

if math.isinf(min_time[y]):
    print(-1)
else:
    print(min_time[y])

File: abc/abc192/e.cpp
#include <iostream>
#include <vector>
#include <array>
#include <queue>

int main() {
    int n, m, x, y;
    std::cin >> n >> m >> x >> y;
    x = x-1;
    y = y-1;
    std::vector<std::vector<int[3]>> path(n);

    for(int i = 0; i < m; ++i) {
        int a, b, t, k;
        std::cin >> a >> b >> t >> k;
        a = a-1;
        b = b-1;
        path[a].push_back({b, t, k});
        path[b].push_back({a, t, k});
    }

    std::vector<int> time_flag(n, -1);
    std::priority_queue<
        std::pair<int, int>,
        std::vector<std::pair<int, int>>,
        std::greater<std::pair<int, int>>> queue;
        queue.push(std::make_pair(0, x));

    while (!queue.empty()) {
        auto [time, town] = queue.top();
        queue.pop();
        if (time_flag[y] != -1 && time_flag[y] < time) {
            break;
        }
        for (const auto& [next, t, k] : path[town]) {
            int next_time = ((time - 1) / k + 1) * k + t;
            if (time_flag[next] == -1 || next_time < time_flag[next]) {
                time_flag[next] = next_time;
                queue.push(std::make_pair(next_time, next));
            }
        }
    }


    std::cout << time_flag[y] << std::endl;
    return 0;
}
